### ISSUE TITLE
Add a transformer like trpc

### ISSUE THREAD

**Issue by mattiavitturi** [inngest/inngest#418]
**Is your feature request related to a problem? Please describe.**
Date objects get converted to string

**Describe the solution you'd like**
Use superjson as a transformer (or any other alternative)

**Describe alternatives you've considered**
For the moment you need to manually convert to superjson and then send the event / parse back on the handler

**Additional context**
some api like `serve(inngest, [fns], { transformer: superjson })`

**Comment by jpwilliams** [comment]
@mattiavitturi Yes! We've discussed this a few times to combat JSON serialization being pretty rubbish; `Date`, as you've found, is usually the first annoyance people run into.

[blitz-js/superjson](https://github.com/blitz-js/superjson) would be preferred due to its continually-growing support in popular JS frameworks, though [MongoDB Extended JSON](https://www.mongodb.com/docs/manual/reference/mongodb-extended-json/) has been another candidate due to its multi-language support for the future.

A user-defined transformer is a great idea. There are a couple of challenges to review:

- Data may arrive in an unexpected format when initially adding/removing the transformer, as the event data and function runs are not necessarily running the same code.

  We could perhaps assume that if the chosen transformer throws or returns `undefined` then we should try the regular `JSON` methods as a backup. This solves adding the transformer, but detecting removal might be trickier.

- Typing for step returns currently uses a [`Jsonfiy<>`](https://github.com/sindresorhus/type-fest/blob/main/source/jsonify.d.ts) type to show that `Date` becomes a `string`, etc. I think the most reasonable approach here is to be lazy and assume that if a transformer is specified then step return will be exactly the same as it would be in-process. This likely won't actually be the case (e.g. a `User` object for a database), but requiring that transformers also provide a type for transformation seems like a stretch.

Would love to get your feedback on the above, @mattiavitturi. 🙂 This is a great suggestion to remove a fair amount of boilerplate.

**Comment by mattiavitturi** [comment]
Sorry for the late reply

Maybe the simplest but not the best solution could be that if you choose to use a transformer then events are valid if and only if the parsing succeed, otherwise they are discarded.

Now that I think about it `trpc-openapi` fully ignores the transformer on input and output and gives to zod the raw json decoded object. 

An option could be to opt-in to transformation on a per-step basis, personally, it's a chore but totally manageable.

Another one is to provide an optional metadata field under the hood, iirc superjson outputs `{ data: {}, meta: {}}` or something similar, but in your case `data` is the real data of the event and `meta` the event metadata for the transformer. This in conjunction with the one above would be safe for steps that don't use a transformer, as it would return `Jsonify<T>` and `T` in case of a transformer.

For our start-up we are using 100% typescript so I can accept to "tie" myself with a transformer.

Being fully optional gives the responsibility of the user to choose wisely and consider the pros/cons

**Comment by jpwilliams** [comment]
@mattiavitturi Thanks for this thoughtful reply!

---

> For our start-up we are using 100% typescript so I can accept to "tie" myself with a transformer.

💯 - This is the case for many folks in the Jamstack/serverless space, so I agree that giving the user the responsibility is reasonable here.

---

> Another one is to provide an optional metadata field under the hood, iirc superjson outputs `{ data: {}, meta: {}}` or something similar, but in your case `data` is the real data of the event and `meta` the event metadata for the transformer.

Yeah, I did want to try leaning into this, but the names feel way too likely to be keys someone would define on their own instead of some `___hidden_dont_use__this` meta key that we could embed silently.

---

Broadly I think, as you say, allowing the user to control this and manage the consequences themselves does seem reasonable. Regarding `Jsonify<T>`, we could use the signature of whatever `deserialize` function is being passed as the `transform` class/object and infer this if it's something explicit (i.e. not `unknown` or `any`) so that users can also customize that in the future.

I will write up an internal spec for this to get feedback from the rest of the team and see what we think. Transforming step data is an easy win as that's all very likely to be in a single application. In contrast, event data might be a future foot gun as those everlasting facts are forever tied to a particular transformer.

**Comment by jpwilliams** [comment]
Transferring this to the SDK repo as it'll likely be a change there.
